runtime/memprof.c

```diff
@@ -334,264 +346,322 @@ struct tracked {
 {
   struct tracked *t;
-  trackst.len++;
-  if (!realloc_trackst()) {
-    trackst.len--;
-    return Invalid_index;
+  if (entries_local == NULL) {
+    entries_local =
+      (struct entry_array*) caml_stat_alloc_noexc(sizeof(struct entry_array));
+    if (entries_local == NULL) return Invalid_index;
+    entries_local->t = NULL;
+    entries_local->min_alloc_len = MIN_ENTRIES_LOCAL_ALLOC_LEN;
+    entries_local->alloc_len = entries_local->len = 0;
+    entries_local->young_idx = entries_local->delete_idx = 0;
   }
-  t = &trackst.entries[trackst.len - 1];
+  if (!realloc_entries(entries_local, 1))
+    return Invalid_index;
+  entries_local->len++;
+  t = &entries_local->t[entries_local->len - 1];
   t->block = block;
   t->n_samples = n_samples;
   t->wosize = wosize;
   t->user_data = user_data;
-  t->idx_ptr = NULL;
   t->alloc_young = is_young;
   t->unmarshalled = is_unmarshalled;
   t->promoted = 0;
   t->deallocated = 0;
-  t->cb_alloc_called = t->cb_promote_called = t->cb_dealloc_called = 0;
+  t->cb_promote_called = t->cb_dealloc_called = 0;
   t->deleted = 0;
   t->callback_running = 0;
-  return trackst.len - 1;
+  return entries_local->len - 1;
 }
 
-static void mark_deleted(uintnat t_idx)
+static void mark_deleted(struct entry_array* ea, uintnat t_idx)
 {
-  struct tracked* t = &trackst.entries[t_idx];
+  struct tracked* t = &ea->t[t_idx];
   t->deleted = 1;
   t->user_data = Val_unit;
   t->block = Val_unit;
-  if (t_idx < trackst.delete) trackst.delete = t_idx;
-  CAMLassert(t->idx_ptr == NULL);
+  if (t_idx < ea->delete_idx) ea->delete_idx = t_idx;
 }
 
-/* The return value is an exception or [Val_unit] iff [*t_idx] is set to
-   [Invalid_index]. In this case, the entry is deleted.
-   Otherwise, the return value is a [Some(...)] block. */
-Caml_inline value run_callback_exn(uintnat *t_idx, value cb, value param) {
-  struct tracked* t = &trackst.entries[*t_idx];
+Caml_inline value run_callback_exn(
+  struct entry_array* ea, uintnat t_idx, value cb, value param)
+{
+  struct tracked* t = &ea->t[t_idx];
   value res;
-  CAMLassert(!t->callback_running && t->idx_ptr == NULL);
+  CAMLassert(!t->callback_running);
   CAMLassert(lambda > 0.);
 
-  callback_running = t->callback_running = 1;
-  t->idx_ptr = t_idx;
+  callback_status = ea == &entries_global ? t_idx : CB_LOCAL;
+  t->callback_running = 1;
+  t->user_data = Val_unit;      /* Release root. */
   res = caml_callback_exn(cb, param);
-  callback_running = 0;
-  /* The call above can modify [*t_idx] and thus invalidate [t]. */
-  if (*t_idx == Invalid_index) {
-    /* Make sure this entry has not been removed by [caml_memprof_set] */
+  if (callback_status == CB_STOPPED) {
+    /* Make sure this entry has not been removed by [caml_memprof_stop] */
     return Val_unit;
   }
-  t = &trackst.entries[*t_idx];
-  t->idx_ptr = NULL;
+  /* The call above can move the tracked entry and thus invalidate
+     [t_idx] and [t]. */
+  if (ea == &entries_global) {
```

Comment:
The following assertion here causes no failures in the testsuite:
```c
CAMLassert(t_idx == callback_status);
```
So I don't think the somewhat tricky invalidation/moving logic is actually tested anywhere.

Reply:
See https://github.com/stedolan/ocaml/commit/0f864c94bd442ae3017a8f508daf27b10929d30d